presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -65,48 +78,221 @@ public CachingCostProvider(CostCalculator costCalculator, StatsProvider statsPro
+    {
+        @Override
+        protected MemoryEstimate visitPlan(PlanNode node, EstimationContext context)
+        {
+            if (context.sourcesCosts.isEmpty()) {
+                return new MemoryEstimate(context.localMaxMemory, context.localMaxMemory);
+            }
+
+            // A PlanNode that is estimated to use memory should have an explicit rule here.
+            verify(isNaN(context.localMaxMemory) || context.localMaxMemory == 0, "Unexpected local memory usage estimation for %s: %s", node, context.localMaxMemory);
+            // A PlanNode that has more than 1 source should have an explicit rule here.
+            double sourceMaxMemoryWhenOutputting = getOnlySourceMaxMemoryWhenOutputting(node, context);
+            return new MemoryEstimate(sourceMaxMemoryWhenOutputting, sourceMaxMemoryWhenOutputting);
+        }
+
+        @Override
+        public MemoryEstimate visitGroupReference(GroupReference node, EstimationContext context)
+        {
+            throw new IllegalStateException("Unexpected GroupReference");
+        }
+
+        @Override
+        public MemoryEstimate visitExchange(ExchangeNode node, EstimationContext context)
+        {
+            PlanCostEstimate sourcesCost = context.sourcesCosts.stream()
+                    .reduce(PlanCostEstimate.zero(), CachingCostProvider::addSiblingsCost);
+
+            return new MemoryEstimate(
+                    sourcesCost.getMaxMemory(),
+                    context.localMaxMemory + sourcesCost.getMaxMemoryWhenOutputting());
+        }
+
+        @Override
+        public MemoryEstimate visitAggregation(AggregationNode node, EstimationContext context)
+        {
+            double sourceMaxMemoryWhenOutputting = getOnlySourceMaxMemoryWhenOutputting(node, context);
+            if (node.getStep() == AggregationNode.Step.SINGLE || node.getStep() == AggregationNode.Step.FINAL) {
+                return new MemoryEstimate(
+                        context.localMaxMemory + sourceMaxMemoryWhenOutputting,
+                        context.localMaxMemory);
+            }
+            else {
+                return new MemoryEstimate(
+                        context.localMaxMemory + sourceMaxMemoryWhenOutputting,
+                        context.localMaxMemory + sourceMaxMemoryWhenOutputting);
+            }
+        }
+
+        @Override
+        public MemoryEstimate visitJoin(JoinNode node, EstimationContext context)
+        {
+            verify(context.sourcesCosts.size() == 2, "Unexpected number of sources for %s: %s", node, context.sourcesCosts.size());
+            double localMaxMemory = context.localMaxMemory;
+            double probeMaxMemoryWhenOutputting = context.sourcesCosts.get(0).getMaxMemoryWhenOutputting();
+            double buildMaxMemoryWhenOutputting = context.sourcesCosts.get(1).getMaxMemoryWhenOutputting();
+
+            return new MemoryEstimate(
+                    localMaxMemory + buildMaxMemoryWhenOutputting,
```

Comment:
why doesn't the beforeOutputting memory estimate include probeMaxMemoryWhenOutputting?  Sometimes the probe side computations will have started while the hash table is being built.  it depends on scheduling, complexity, etc. whether it uses memory before hitting the buffer limit.

Reply:
Good catch